libafl_nyx/src/executor.rs

```diff
@@ -0,0 +1,98 @@
+use std::fmt::Debug;
+/// executor for nyx
+use std::marker::PhantomData;
+
+use libafl::{
+    executors::{Executor, ExitKind, HasObservers},
+    inputs::{HasBytesVec, Input},
+    observers::ObserversTuple,
+    Error,
+};
+use libnyx::NyxReturnValue;
+
+use crate::helper::NyxHelper;
+
+/// executor for nyx standalone mode
+pub struct NyxExecutor<'a, I, S, OT> {
+    /// implement nyx function
+    pub helper: &'a mut NyxHelper,
+    /// observers
+    observers: OT,
+    /// phantom data to keep generic type <I,S>
+    phantom: PhantomData<(I, S)>,
+}
+
+impl<'a, I, S, OT> Debug for NyxExecutor<'a, I, S, OT> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NyxInprocessExecutor")
+            .field("helper", &self.helper)
+            .finish()
+    }
+}
+
+impl<'a, EM, I, S, Z, OT> Executor<EM, I, S, Z> for NyxExecutor<'a, I, S, OT>
+where
+    I: Input + HasBytesVec,
+{
+    fn run_target(
+        &mut self,
+        _fuzzer: &mut Z,
+        _state: &mut S,
+        _mgr: &mut EM,
+        input: &I,
+    ) -> Result<libafl::executors::ExitKind, libafl::Error> {
+        let input = input.bytes();
+        self.helper.nyx_process.set_input(input, input.len() as u32);
+
+        // exec will take care of trace_bits, so no need to reset
+        let ret_val = self.helper.nyx_process.exec();
+        match ret_val {
+            NyxReturnValue::Normal => Ok(ExitKind::Ok),
+            NyxReturnValue::Crash | NyxReturnValue::Asan => Ok(ExitKind::Crash),
+            NyxReturnValue::Timeout => Ok(ExitKind::Timeout),
+            NyxReturnValue::InvalidWriteToPayload => Err(libafl::Error::illegal_state(
+                "FixMe: Nyx InvalidWriteToPayload handler is missing",
+            )),
+            NyxReturnValue::Error => Err(libafl::Error::illegal_state(
+                "Error: Nyx runtime error has occured...",
+            )),
+            NyxReturnValue::IoError => {
+                // todo! *stop_soon_p = 0
+                Err(libafl::Error::illegal_state("Error: QEMU-nyx died..."))
+            }
+            NyxReturnValue::Abort => {
+                self.helper.nyx_process.shutdown();
+                Err(libafl::Error::illegal_state("Error: Nyx abort occured..."))
+            }
+        }
+    }
+}
+
+impl<'a, I, S, OT> NyxExecutor<'a, I, S, OT> {
+    pub fn new(helper: &'a mut NyxHelper, observers: OT) -> Result<Self, Error> {
```

Comment:
Just for your consideration, WRT API design:
You could do a Builder for Nyx, and call `.build(observers)` on it in the end to create the NyxExecutor, instead of passing in a NyxHelper to `new`.
Alternatively, you could just call `build` orn the Helper (and maybe rename it)

Reply:
Thanks for the suggestion! But I think that currently the code scale for NYX is still small, so a single builder may not bring lots of benefits. And adding a method like `Helper.build()` involves adding several generic types to the `helper`. So maybe we can wait for further development.

Reply:
You could add the generic types to the `build` function, look at  this example:
https://github.com/AFLplusplus/LibAFL/blob/7d7601204f501959cefa3654e5e2ef82b44a3584/libafl/src/executors/command.rs#L535

But we can do it also for the next step, I agree

Reply:
Also look at the crate we use for the builder here, it's super easy to do (later)

Reply:
yes we want builder pattern when we have lots of options for the nyxfuzzer in the future (Redqueen mode, timeout, ... etc)